zerver/tests/test_messages.py

```diff
@@ -2265,3 +2266,109 @@ def log_change_in_subscription(user, stream, event_type):
+
+    def test_user_message_filter(self):
+        # type: () -> None
+        # In this test we are basically testing out the logic used out in
+        # do_send_messages() in action.py for filtering the messages for which
+        # UserMessage rows should be created for a soft-deactivated user.
+        recipient_list  = [
+            self.example_email("hamlet"),
+            self.example_email("iago"),
+            self.example_email('cordelia')
+        ]
+        for email in recipient_list:
+            self.subscribe_to_stream(email, "Denmark")
+
+        cordelia = self.example_user('cordelia')
+        sender = self.example_email('iago')
+        stream_name = 'Denmark'
+        subject = 'foo'
+
+        def send_stream_message(content):
+            # type: (str) -> None
+            self.send_message(sender, stream_name, Recipient.STREAM,
+                              content, subject)
+
+        def send_personal_message(content):
+            # type: (str) -> None
+            self.send_message(sender, self.example_email("hamlet"),
+                              Recipient.PERSONAL, content)
+
+        long_term_idle_user = self.soft_deactivate_user('hamlet')
+
+        def assert_um_count(user, count):
+            # type: (UserProfile, int) -> None
+            user_messages = self.get_user_ums(user)
+            self.assertEqual(len(user_messages), count)
+
+        def assert_last_um_content(user, content, negate=False):
+            # type: (UserProfile, Text, bool) -> None
+            user_messages = self.get_user_ums(user)
+            if negate:
+                self.assertNotEqual(user_messages[-1].message.content, content)
+            else:
+                self.assertEqual(user_messages[-1].message.content, content)
+
+        # Test that sending a message to a stream with soft deactivated user
+        # doesn't end up creating UserMessage row for deactivated user.
+        general_user_msg_count = len(self.get_user_ums(cordelia))
+        soft_deactivated_user_msg_count = len(self.get_user_ums(long_term_idle_user))
+        message = 'Test Message 1'
+        send_stream_message(message)
+        assert_last_um_content(long_term_idle_user, force_text(message), negate=True)
+        assert_um_count(long_term_idle_user, soft_deactivated_user_msg_count)
+        assert_um_count(cordelia, general_user_msg_count + 1)
+        assert_last_um_content(cordelia, force_text(message))
+
+        # Test sending a private message to soft deactivated user creates
+        # UserMessage row.
+        soft_deactivated_user_msg_count = len(self.get_user_ums(long_term_idle_user))
+        message = 'Test PM'
+        send_personal_message(message)
+        assert_um_count(long_term_idle_user, soft_deactivated_user_msg_count + 1)
+        assert_last_um_content(long_term_idle_user, force_text(message))
+
+        # Test UserMessage row is only created while user is deactivated if
+        # user itself is mentioned.
```

Comment:
Yup, post the `um.flags == 0` check upgrade, I have added new tests for the other cases as well including this one.